File: panel/backend/internal/models/activity.go
package models

import "time"

// ActivityCategory is the broad bucket the Activity page groups by.
type ActivityCategory string

const (
	ActivityCategoryUser     ActivityCategory = "user"
	ActivityCategoryRole     ActivityCategory = "role"
	ActivityCategoryNode     ActivityCategory = "node"
	ActivityCategoryTemplate ActivityCategory = "template"
	ActivityCategoryInstance ActivityCategory = "instance"
	ActivityCategoryAPIKey   ActivityCategory = "api_key"
	ActivityCategorySettings ActivityCategory = "settings"
	ActivityCategoryAuth     ActivityCategory = "auth"
	ActivityCategorySystem   ActivityCategory = "system"
	ActivityCategoryMod      ActivityCategory = "mod"
	ActivityCategorySecret   ActivityCategory = "secret"
	ActivityCategoryAutomationCategory ActivityCategory = "automation"
	ActivityCategorySnapshot ActivityCategory = "snapshot"
	ActivityCategorySecurity ActivityCategory = "security"
)

// ActivityLog is one row of the admin audit timeline.
//
// UserID / UserIDValid model the "deleted user" case — a historically valid
// user_id whose users row is gone should still render the activity row, with
// the denormalised `Username` field carrying the label.
type ActivityLog struct {
	ID          int64            `json:"id"`
	UserID      *int64           `json:"user_id,omitempty"`
	Username    string           `json:"username"`
	Role        string           `json:"role"`
	Category    ActivityCategory `json:"category"`
	Action      string           `json:"action"`
	TargetID    *int64           `json:"target_id,omitempty"`
	TargetLabel string           `json:"target_label"`
	Message     string           `json:"message"`
	IPAddress   string           `json:"ip_address"`
	UserAgent   string           `json:"user_agent"`
	CreatedAt   time.Time        `json:"created_at"`
}

File: panel/frontend/src/features/activity/types/activity.ts
// ActivityLog is one row of the admin audit timeline rendered by the
// Activity page. The shape is intentionally rich: each field gets its own
// dedicated card element, so the page never has to parse a free-form
// message string to figure out "who / what / when".

export type ActivityCategory =
  | 'user'
  | 'role'
  | 'node'
  | 'template'
  | 'instance'
  | 'api_key'
  | 'settings'
  | 'auth'
  | 'system';

export interface ActivityLog {
  id: number;
  // Nullable when the actor's users row was deleted after the action.
  user_id?: number | null;
  // Denormalised at write-time so the row stays readable post-delete.
  username: string;
  // Display name of the actor's role at the time of the action.
  role?: string;
  category: ActivityCategory;
  // Action verb: 'create' | 'update' | 'delete' | 'login' | 'login_failed'
  // | 'rotate_token' | 'probe' | 'deploy' | 'start' | 'stop' | 'destroy'.
  action: string;
  target_id?: number | null;
  // Human-readable label of the affected entity.
  target_label?: string;
  message: string;
  // Client IP captured at write-time. X-Forwarded-For friendly.
  ip_address: string;
  // Raw User-Agent header value (truncated).
  user_agent: string;
  created_at: string;
}
